service/frontend/service.go

```diff
@@ -334,19 +334,18 @@ func (s *Service) Stop() {
 
 	// initiate graceful shutdown:
 	// 1. Fail rpc health check, this will cause client side load balancer to stop forwarding requests to this node
-	// 2. wait for failure detection time
+	// 2. wait for 10 seconds failure detection time
 	// 3. stop taking new requests by returning InternalServiceError
-	// 4. Wait for a second
+	// 4. Wait for X second
 	// 5. Stop everything forcefully and return
 
-	requestDrainTime := util.Min(time.Second, s.config.ShutdownDrainDuration())
-	failureDetectionTime := util.Max(0, s.config.ShutdownDrainDuration()-requestDrainTime)
+	requestDrainTime := util.Max(time.Second, s.config.ShutdownDrainDuration())
 
 	logger.Info("ShutdownHandler: Updating gRPC health status to ShuttingDown")
 	s.healthServer.Shutdown()
 
 	logger.Info("ShutdownHandler: Waiting for others to discover I am unhealthy")
-	time.Sleep(failureDetectionTime)
+	time.Sleep(10 * time.Second)
```

Comment:
should we have another dynamic config for this? this actually seems like the one that's more dependent on the environment (external health check frequency). the requestDrainTime can be fixed to 5s or 10s since we use 5s or 10s timeout on rpcs

Reply:
+1 for having a separate knob for this.

The timeout is from client which can be quite long I think? Or may be I misunderstood something?

Reply:
sure. Will add a different config.

Reply:
My understanding is during this sleep, any rpcs that end up here will still be handled, but we expect some external system to do a health check, notice the "shutting down" response, and adjust its state to stop sending rpcs here. That timeout is controlled by the load balancer.

The second sleep is when we stop accepting rpcs, but continue processing ones that have already come in. That one depends on how long we expect our operations to take, which we have more control over. For long-polls, we can just fail and let them get retried. For everything else, if it takes more than 10s something is probably going wrong, so it seems okay to fail. But no harm in making that configurable too